Wrappers/Python/ccpi/optimisation/functions/MixedL21Norm.py

```diff
@@ -182,9 +145,68 @@ def proximal(self, x, tau, out=None):
 #            res1.maximum(1.0, out=res1)	
 #            x.divide(res1, out=out)
                               
+                
+class SmoothMixedL21Norm(Function):
+    
+    """ SmoothMixedL21Norm function: :math:`F(x) = ||x||_{2,1} = \sum |x|_{2} = \sum \sqrt{ (x^{1})^{2} + (x^{2})^{2} + \epsilon^2 + \dots}`                  
+    
+        where x is a BlockDataContainer, i.e., :math:`x=(x^{1}, x^{2}, \dots)`
+        
+        Conjugate, proximal and proximal conjugate methods no closed-form solution
+        
+    
+    """    
+        
+    def __init__(self, epsilon):
+                
+        r'''
+        :param epsilon: smoothing parameter making MixedL21Norm differentiable 
+        '''
+
+        super(SmoothMixedL21Norm, self).__init__(L=1)          
+        self.epsilon = epsilon   
+                
+        if self.epsilon==0:
+            raise ValueError('We need epsilon>0. Otherwise, call "MixedL21Norm" ')
+                            
+    def __call__(self, x):
+        
+        r"""Returns the value of the SmoothMixedL21Norm function at x.                                            
+        """
+        if not isinstance(x, BlockDataContainer):
+            raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(x))) 
+            
+        tmp = x.copy()            
+        tmp.containers += (self.epsilon,)     
+        return tmp.pnorm(p=2).sum()             
+         
+
+    def gradient(self, x, out=None): 
+        
+        r"""Returns the value of the gradient of the SmoothMixedL21Norm function at x.
+        
+        \frac{x}{|x|}
+                
+                
+        """     
+        
+        if not isinstance(x, BlockDataContainer):
+            raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(x))) 
+                   
+        
+        tmp = x.copy()
+        tmp.containers += (self.epsilon,)          
```

Comment:
I think that 
```python

tmp = x.add(self.epsilon)
```
should do what you require.

Reply:
Prefer, as I have it, so it is a tuple in the form `(x1, x2, epsilon)` and then use the `pnorm(2)` and do `(x1^2, x2^2, epsilon^2)` and take the sum.

Reply:
I think that this reduces the readibility of the code. It is not obvious that that is what you want to achieve. 
If I understand correctly, you could achieve that by
```python
denom = x.pnorm(2) + self.epsilon**2
```
which is way more explicit and uses the `pnorm` in a canonical way.

Reply:
The above denom is actually `np.sqrt( x1^2 + x2^2) + epsilon^2` different from `x.pnorm(2)` when `x=(x1,x2, epsilon)`